### PR TITLE
Review parser - parse restaurant review articles by Marina O'Loughlin.

### PR DESCRIPTION
**Pull request by LATaylor-guardian**
This PR attempts to parse restaurant reviews for a given reviewer. The idea is to write a different strategy for each reviewer (using commonalities between them where we can) given the difference in styles for each review per the reviewer. As such, I have tried to structure the code in such a way that makes it flexible to do so. 

`MarinaOLoughlinReviewParser.scala` provides the bulk of the parsing logic - this could still do with some work to make it more readable while highlighting the rules being applied/assumptions taken for given data. I will do so in a subsequent PR and pull out much of the parsing logic into extractor objects. 

The code should be sufficiently under test. When we encounter any edge cases/new examples of data that doesn't parse we should add a test and then address parsing them correctly.

Apologies for the large PR, what went from hacking around to see if I could parse restaurant reviews turned into something more.

### CODE REVIEW

File: src/main/scala/com/gu/restaurant_review_parser/parsers/MarinaOLoughlinReviewParser.scala [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  def guessRestaurantWebAddress(articleBody: ArticleBody, restaurantName: RestaurantName): Option[WebAddress] = {
+    val doc = Jsoup.parse(articleBody.value)
+
+    val maybeAddress = doc.select("p > a").iterator().asScala.toList
+      .filter(e => e.text.toLowerCase.contains(restaurantName.value.toLowerCase)).lastOption // assume the one we want is always the last one.
+
+    maybeAddress.map(elem => WebAddress(elem.attr("href")))
+  }
+
+  def guesstRatingBreakdown(articleBody: ArticleBody): Option[RatingBreakdown] = {
```
Comment (tomrf1) [inline src/main/scala/com/gu/restaurant_review_parser/parsers/MarinaOLoughlinReviewParser.scala]: typo in the name?

Comment (LATaylor-guardian) [inline src/main/scala/com/gu/restaurant_review_parser/parsers/MarinaOLoughlinReviewParser.scala]: Yeah, good spot.

File: src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+case class RatingBreakdown(food: FoodRating, atmosphere: AtmosphereRating, valueForMoney: ValueForMoneyRating)
+
+case class ParsedRestaurantReview (
+  restaurantName: RestaurantName,
+  approximateLocation: ApproximateLocation,
+  reviewer: String,
+  publicationDate: OffsetDateTime,
+  ratingBreakdown: Option[RatingBreakdown],
+  address: Option[Address], // TODO - structure accordingly
+  webAddress: Option[WebAddress]
```
Comment (tsop14) [inline src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala]: Do we want to include phone number and opening hours? Or will these go under address?

**Comment by tsop14** [comment]
Nice!

File: src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
+package com.gu.restaurant_review_parser
+
+import java.time.OffsetDateTime
+
+case class WebAddress(value: String) extends AnyVal
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala]: They're value classes (http://docs.scala-lang.org/overviews/core/value-classes.html) so that we have some type safety around all of the Strings being passed around.

File: src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala (hunk 2 above)
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala]: The address field is currently just the block of text that contains the address along with other useful information. More work needs to be done to further extract structure from this. This is to come! :)

File: src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala (hunk 3 above)
Comment (oilnam) [inline src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala]: @tsop14 check out this guy I know called Chris: http://slides.com/cb372/real-world-func-scala-lambdaworld-2016#/15 ;)

Comment (LATaylor-guardian) [inline src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala]: Heard he's a great GO developer.

Comment (mchv) [inline src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala]: The slides (@cb372) and the code is missing the mandatory `val` to be value classes.
> which is marked with val and which has public accessibility

Comment (LATaylor-guardian) [inline src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala]: Good spot.

Comment (LATaylor-guardian) [inline src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala]: @mchv Actually surely this isn't required? It's a case class not a class?

Comment (LATaylor-guardian) [inline src/main/scala/com/gu/restaurant_review_parser/ParsedRestaurantReview.scala]: GHOST!

File: src/main/scala/com/gu/restaurant_review_parser/Reviewers.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  val body: Option[String]
+  val webPublicationDate: Option[CapiDateTime]
+}
+
+object ReviewArticle {
+  def apply(reviewer: Reviewer, content: Content): ReviewArticle = {
+    val id = content.id
+    val webTitle = content.webTitle
+    val byline = content.fields.flatMap(_.byline)
+    val body = content.fields.flatMap(_.body)
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/restaurant_review_parser/Reviewers.scala]: We rely on the html at times to parse it.
